rust/arrow/src/array/data.rs

```diff
@@ -225,11 +226,11 @@ pub struct ArrayData {
     /// The buffers for this array data. Note that depending on the array types, this
     /// could hold different kinds of buffers (e.g., value buffer, value offset buffer)
     /// at different positions.
-    buffers: Vec<Buffer>,
+    buffers: SmallVec<[Buffer; 1]>,
```

Comment:
I wonder if using something like a specific enum might also work (and be more memory efficient) as well:

```
pub enum  ArrayBuffer {
  One(ArrayDataRef),
  Two(ArrayDataRef, ArrayDataRef),
  Many(Vec<ArrayDataRef>)
}
```

I honestly don't know what the maximum number of buffers any specific ArrayData can have,

Reply:
fyi, max is 3 including the null buffer, so max is 2 in this context. See `MutableDataArray`, where we use two buffers on the stack to avoid the overhead. When unused, a buffer takes something like a `usize` + a pointer.

Reply:
Good idea! I haven't checked the memory layout of both, but in the common case of 1 buffer this could inline better (in the absence of LTO). Since we already have the validity bitmap in a separate Option, the variants are `Zero`, `One` and `Many`.